contracts/staking/GoodCompoundStaking.sol

```diff
@@ -4,91 +4,125 @@ pragma solidity >=0.7.0;
-        
-    }
-
-    /**
-     * @dev stake some DAI
-     * @param _amount of dai to stake
-     */
-    function mintInterestToken(uint256 _amount) internal override{
-        
-        cERC20 cToken = cERC20(address(iToken));
-        uint res = cToken.mint(_amount);
-
-        if (
-            res > 0
-        ) //cDAI returns >0 if error happened while minting. make sure no errors, if error return DAI funds
-        {
-            require(res == 0, "Minting cDai failed, funds returned");
-        }
-
-    }
-
-    /**
-     * @dev redeem DAI from compound 
-     * @param _amount of dai to redeem
-     */
-    function redeem(uint256 _amount) internal override{
-        cERC20 cToken = cERC20(address(iToken));
-        require(cToken.redeemUnderlying(_amount) == 0, "Failed to redeem cDai");
-
-    }
-
-    /**
-     * @dev returns Dai to cDai Exchange rate.
-     */
-    function exchangeRate() internal view override returns(uint) {
-        cERC20 cToken = cERC20(address(iToken));
-        return cToken.exchangeRateStored();
-
-    }
-
-    /**
-     * @dev returns decimals of token.
-     */
-    function tokenDecimal() internal view override returns(uint) {
-        ERC20 token = ERC20(address(token));
-        return uint(token.decimals());
-    }
-
-    /**
-     * @dev returns decimals of interest token.
-     */
-    function iTokenDecimal() internal view override returns(uint) {
-        ERC20 cToken = ERC20(address(iToken));
-        return uint(cToken.decimals());
-    }
-
-    function getGasCostForInterestTransfer() external view override returns(uint256){
-        return uint256(67917);
-    }
-}
\ No newline at end of file
+	/**
+	 * @param _token Token to swap DEFI token
+	 * @param _iToken DEFI token address
+	 * @param _ns Address of the NameService
+	 * @param _tokenName Name of the staking token which will be provided to staker for their staking share
+	 * @param _tokenSymbol Symbol of the staking token which will be provided to staker for their staking share
+	 * @param _tokenSymbol Determines blocks to pass for 1x Multiplier
+	 */
+	constructor(
+		address _token,
+		address _iToken,
+		uint256 _blockInterval,
+		NameService _ns,
+		string memory _tokenName,
+		string memory _tokenSymbol,
+		uint64 _maxRewardThreshold
+	)
+		SimpleStaking(
+			_token,
+			_iToken,
+			_blockInterval,
+			_ns,
+			_tokenName,
+			_tokenSymbol,
+			_maxRewardThreshold
+		)
+	{}
+
+	/**
+	 * @dev stake some DAI
+	 * @param _amount of dai to stake
+	 */
+	function mintInterestToken(uint256 _amount) internal override {
+		cERC20 cToken = cERC20(address(iToken));
+		uint256 res = cToken.mint(_amount);
+
+		if (
+			res > 0
+		) //cDAI returns >0 if error happened while minting. make sure no errors, if error return DAI funds
+		{
+			require(res == 0, "Minting cDai failed, funds returned");
+		}
+	}
+
+	/**
+	 * @dev redeem DAI from compound
+	 * @param _amount of dai to redeem
+	 */
+	function redeem(uint256 _amount) internal override {
```

Comment:
this is called redeem but calls redeemunderlying? and vice versa with the following method

Reply:
yes because cToken.redeemUnderlying works with token amount so we need to use this one in this redeem.
but cToken.redeem work with itoken amount and gives token amount and this redeemunderlying method for
itoken -> DAI transactions in order to collect interest in DAI 
Maybe I should put better doc but stuff is correct there

Reply:
there's a confusion here, GoodCompoundStaking is a generic compound contract not the "SimpleDAIStaking" we had.

all staking contract should have one function that redeem iToken -> DAI
because it has to transfer DAI to staking contract.
this should probably be reflected in the method name

Reply:
Ok I refactored GoodCompoundStaking to be more generic contract